### PR TITLE
feat(python): Add support for Iceberg

### LINKED ISSUE #6227: Add Read support for Apache Iceberg
### Problem description

Would like to see read support for Apache Iceberg similar to the support for delta. Cc @asheeshgarg

### PR DESCRIPTION
**Pull request by Fokko**
Resolves #6227

Using PyIceberg:

- Implements field-id-based schema resolution
- Integration with [Catalogs](https://py.iceberg.apache.org/configuration/), among others, Hive, Glue, DynamoDB, REST, and SqlCatalog.
- Both partition and metrics-based data file pruning; will only load files that are relevant to the expression. Reducing the data being read and the number of calls to the object store.
- Support for AWS, Azure, and GCS is in the next release.
- Using the PyArrow backend.

### CODE REVIEW

**Comment by stinodego** [comment]
Hi @Fokko, looks like you've done quite some work here! Could you please make sure the CI passes? Then we'll take a closer look.

Let us know if you need any help.

**Comment by Fokko** [comment]
@stinodego Thanks! 

> Could you please make sure the CI passes? Then we'll take a closer look.

Yes, I'm on top of it. PyIceberg is still at Pydantic <2. I have a [PR ready on the PyIceberg](https://github.com/apache/iceberg/pull/7782) side to bump it to 2, then we need to do a release, and then it should pass the CI.

**Comment by Fokko** [comment]
@stinodego We're releasing PyIceberg 0.5.0 which also supports Pydantic 2, same as Polars. Would be great to get some eyes on this, I'm quite excited about Iceberg integration in Polars.

**Comment by Fokko** [comment]
@stinodego all green :)

File: py-polars/tests/unit/io/files/iceberg-table/data/tpep_pickup_datetime_day=2022-03-01/00000-129-fd34af88-9f14-4306-908d-91866462ae81-00001.parquet
Comment (stinodego) [inline py-polars/tests/unit/io/files/iceberg-table/data/tpep_pickup_datetime_day=2022-03-01/00000-129-fd34af88-9f14-4306-908d-91866462ae81-00001.parquet:1]: These binary files are quite large - way larger than other data files in the repository. Is it possible to cut these down to a few kilobytes?

File: py-polars/tests/unit/test_polars_import.py
Comment (stinodego) [inline py-polars/tests/unit/test_polars_import.py:70]: This is not acceptable - if you're noticing an impact on import time, it means the implementation does not utilize lazy loading correctly. This change should be reversed.

File: py-polars/polars/io/iceberg.py
Comment (stinodego) [inline py-polars/polars/io/iceberg.py:112]: This mentions Delta tables - must be a copy paste mistake? Please double-check the docstring to make sure it's correct. Example below also mentions `delta_table_options`.

File: py-polars/polars/io/iceberg.py
Comment (stinodego) [inline py-polars/polars/io/iceberg.py:137]: Could you "newspaper" this and define it below the function that uses it?

File: py-polars/polars/io/iceberg.py
Comment (stinodego) [inline py-polars/polars/io/iceberg.py:276]: This is not used anywhere - can be deleted.

File: py-polars/polars/io/iceberg.py
Comment (stinodego) [inline py-polars/polars/io/iceberg.py:186]: Imports should be at the top of the file if possible.

File: py-polars/tests/unit/io/test_iceberg.py
Comment (stinodego) [inline py-polars/tests/unit/io/test_iceberg.py:28]: Testing the length is OK, but definitely test the schema too (at least in some of the tests).

**Review by stinodego — CHANGES_REQUESTED** [review]
Hi @Fokko , I went through the PR and left a few comments. Overall things look good but there are some things to improve.

Will be happy to merge this once all looks good :+1:

File: py-polars/tests/unit/test_polars_import.py
Comment (Fokko) [reply py-polars/tests/unit/test_polars_import.py]: I wasn't aware of this, let me fix this. I've reverted the change.

Comment (Fokko) [reply py-polars/tests/unit/test_polars_import.py]: I have to dig deeper into the lazy loading tomorrow

Comment (stinodego) [reply py-polars/tests/unit/test_polars_import.py]: I can also take a look if you want.

Comment (Fokko) [reply py-polars/tests/unit/test_polars_import.py]: Yes that would be nice, it is not directly obvious to me. It feels like I have to import from `polars.dependencies`, but that doesn't seem to work.

File: crates/polars-plan/src/logical_plan/pyarrow.rs
Comment (stinodego) [reply crates/polars-plan/src/logical_plan/pyarrow.rs]: Very welcome to add these in a new PR, right away or after this has been merged 👍

File: py-polars/tests/unit/test_polars_import.py
Comment (alexander-beedie) [reply py-polars/tests/unit/test_polars_import.py]: > Yes that would be nice, it is not directly obvious to me. It feels like I have to import from `polars.dependencies`, but that doesn't seem to work.

Took a quick look; you had registered `pyiceberg` with the lazy loader, but then imported _directly_ from it at the head of the module, so it was promptly instantiated "for real". Looks fine now though - @stinodego beat me to it, heh ;)

File: py-polars/pyproject.toml
Comment (Fokko) [inline py-polars/pyproject.toml:52]: @stinodego @alexander-beedie All tests are passing. Can we hold off on this PR until 0.5.0 has been released?

Comment (alexander-beedie) [reply py-polars/pyproject.toml:52]: Don't see why not, as long as you don't mind rebasing if/when necessary, heh 😄
Got a rough ETA on the `0.5.0` release? (Presumably not that far away if it's already at `rc3`).

File: py-polars/polars/io/iceberg.py
Comment (stinodego) [inline py-polars/polars/io/iceberg.py:132]: Why do we need a second `_scan_pyarrow_dataset_impl`? Can we not reuse the existing implementation?

**Review by stinodego — APPROVED** [review]
Good stuff. I'm on holiday next week. Feel free to merge once pyiceberg 0.5.0 has been released and the PR has been updated accordingly.

File: py-polars/pyproject.toml
Comment (Fokko) [reply py-polars/pyproject.toml:52]: ETA is roughly around early next week

File: py-polars/polars/io/iceberg.py
Comment (stinodego) [reply py-polars/polars/io/iceberg.py:132]: @Fokko Could you come back to me on this one?

Comment (Fokko) [reply py-polars/polars/io/iceberg.py:132]: Of course. Here we do something different than in the original one. In the original, the Python `eval` function is used to convert the string containing Python to an actual Python class, and that's being passed into the delta library:

```
    if predicate:
        # imports are used by inline python evaluated by `eval`
        from polars.datatypes import Date, Datetime, Duration  # noqa: F401
        from polars.utils.convert import (
            _to_python_datetime,  # noqa: F401
            _to_python_time,  # noqa: F401
            _to_python_timedelta,  # noqa: F401
        )
    
        _filter = eval(predicate)
```

What we do here is that we take the string, we convert it into an abstract syntax tree, and that's being traversed to convert it into a PyIceberg expression. The reason why I did this is that the PyArrow expression doesn't have any Python methods to traverse the expression (the same goes for the Polars expression, otherwise I could just traverse that one as well). I've added this to the docstring as well 👍🏻

**Comment by alexander-beedie** [comment]
Alright... here we go ;)

**Review by alexander-beedie — APPROVED** [review]

**Comment by Fokko** [comment]
Awesome, thanks all! 🚀

**Comment by alexander-beedie** [comment]
> Awesome, thanks all! 🚀

It's a great addition; looking forward to seeing it in action 👍

**Comment by Fokko** [comment]
I did a small demo at PyData Amsterdam last week. Once it is released, I'll work on a blogpost! 👍🏻

**Comment by luancaarvalho** [comment]
@Fokko Do you have the demo at PyData Amsterdam recorded ?

**Comment by Fokko** [comment]
@luancaarvalho The demo is recorded, but not yet online. I think they still have to do some final editing.

**Pull request merged**
